## Getting Started

```bash

# Clone the repository (assuming you place Main.sh in a repo)
git clone https://github.com/junhaiqi/TopoRepeat.git
cd TopoRepeat

# Ensure dependencies (r2rtr, srfutils.js, minimap2) are compiled and placed in their respective directories, one-click installation:
chmod +x install.sh && ./install.sh
# Example: Run the full pipeline on raw reads
bash Main.sh -i testdata/test.fa -o testdata/test_TopoRepeat
```


## Overview of TopoRepeat
`TopoRepeat` is an integrated shell workflow designed to automate the discovery and quantification of tandem repeats (TR) from long-read sequencing data (PacBio HiFi/CLR/ONT). It seamlessly combines read subsampling, self-alignment, unit inference, and sequence clustering into a single execution command.

The pipeline integrates [minimap2](https://github.com/lh3/minimap2) for alignment, [r2rtr](https://github.com/junhaiqi/r2rtr) for unit inference, [raEDClust](https://github.com/junhaiqi/raEDClust) for unit clustering, and [srfutils.js](https://github.com/lh3/srf) for abundance estimation. It allows users to go from raw FASTQ files to clustered consensus units and abundance tables without manual intervention.
## Table of contents

  * [Requirements](#requirements)
  * [Installation](#installation)
  * [Usage](#usage)
  * [Example](#example)
  * [Output](#output)
  * [Acknowledgments](#acknowledgments)
  * [License](#license)
  * [Cite](#cite)

## Requirements
The pipeline runs on Linux and requires the following dependencies:
Python 3.x, pyfastx package (for subsampling)
Minimap2 (in system PATH or specified directory)
GCC 9.3.0+ (for compiling C++ tools)
srfutils.js requires the k8 javascript engine


## Installation
### One-Click installation:
```bash
chmod +x install.sh && ./install.sh
```
### Install step by step:

1.Install `minimap2`
```bash
cd minimap2 && make -j8
```
2.Install `r2rtr`
```bash
cd r2rtr && make -j8
```
3.Install `raEDClust`
```bash
cd raEDClust && make -j8
```
4.Install `sample_reads.py`
```bash
pip install pyfastx
```
5.Install `srfutils.js`
This script *requires* the [k8 Javascript shell][k8] to run. On
Linux or Mac, you can download the precompiled k8 binary with:

```sh
curl -L https://github.com/attractivechaos/k8/releases/download/v0.2.4/k8-0.2.4.tar.bz2 | tar -jxf -
cp k8-0.2.4/k8-`uname -s` $HOME/bin/k8  # assuming $HOME/bin in your $PATH
```

It is highly recommended to copy the executable `k8` to a directory on your
`$PATH` such as `/usr/bin/env` can find it. Like python scripts, once you
install `k8`, you can launch srfutils.js in one of the two ways:

```sh
path/to/srfutils.js             # only if k8 is on your $PATH
k8 path/to/srfutils.js
```


## Usage

Basic command:
```bash
./Main.sh -i <input_fastq> -o <output_dir> [Options]
```
The specific parameters are as follows:
```bash
Required:
  -i <path>   Input FASTQ file (raw reads, supports .gz)
  -o <path>   Output directory path

Options [Global]:
  -t <int>    Number of threads to use [default = 16]

Options [Step 1: Sampling & Alignment]:
  -p <float>  Sampling percentage (0-100) [default = 10]
  -x <str>    Minimap2 preset (ava-pb or ava-ont) [default = ava-pb]
  -L <int>    Minimum read length to retain during sampling [default = 10000]

Options [Step 2: r2rtr Inference]:
  -n <int>    r2rtr: Minimum number of supports for the TR length [default = 10]
  -k <int>    r2rtr: Minimum inferred tandem repeat length [default = 10]
  -c <0|1>    r2rtr: Accurate reads mode (0=Noisy, 1=HiFi/ONT-R10) [default = 0]

Options [Step 3: raClust-SRF Analysis]:
  -S <float>  Clustering similarity threshold (raEDClust) [default = 0.90]
  -R <float>  Clustering length ratio (raEDClust) [default = 0.85]
  -h          Show help message
```

## Example

Run the pipeline on PacBio CLR data, sampling 10% of reads, using 24 threads, and stricter clustering parameters:

```bash
bash Main.sh \
  -i testdata/test.fa \
  -o testdata/test_TopoRepeat \
  -t 24 \
  -p 10 \
  -x ava-pb \
  -S 0.95
```

The command line above estimates the abundance of repeat units by mapping them to raw reads. If you already have repeat units (units.fa) and a high-quality reference genome or contigs (ref.fa), you can estimate their abundance based on those:
```bash
bash TRAnalyzer.sh -i ref.fa -r units.fa -o output
```

## Output
The output directory will contain three subdirectories corresponding to the pipeline steps:
1. 01_Sampling/
* *_10pct.fq.gz: The subsampled reads.
* *_10pct.paf: Self-alignment file generated by Minimap2.
2. 02_r2rtr/, details in [r2rtr](https://github.com/junhaiqi/r2rtr)
* *_r2rtr_units.fasta: Raw tandem repeat units inferred by r2rtr.
3. 03_raEDClust_SRF/ (Final Results)
* `raEDClust` results, details in [raEDClust](https://github.com/junhaiqi/raEDClust)
    * *.clustered.fa: The final clustered consensus sequences of the repeat units.
    * *.clusters.txt: Details of which raw units belong to which cluster.

* `srfutils.js` results, details in [srf](https://github.com/lh3/srf)
    * *_abundance.txt: The final abundance table showing the count/coverage of each repeat cluster in the sampled data.
    * *.mapping.paf: The results of `minimap2` by command `minimap2 -c -N1000000 -f1000 -r100,100 <(./srfutils.js enlong r2rtr.fa) ref.fa > *.mapping.paf`
    * *.bed : The results of `./srfutils.js` by command `paf2bed *.mapping.paf > *.bed`


## License 
MIT License.

## Cite
None





